shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -0,0 +1,524 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine;
+
+import android.content.res.AssetManager;
+import android.graphics.Bitmap;
+import android.graphics.SurfaceTexture;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.UiThread;
+import android.view.Surface;
+
+import java.nio.ByteBuffer;
+import java.util.HashSet;
+import java.util.Set;
+
+import io.flutter.embedding.engine.dart.PlatformMessageHandler;
+import io.flutter.embedding.engine.FlutterEngine.EngineLifecycleListener;
+import io.flutter.embedding.engine.renderer.FlutterRenderer;
+import io.flutter.embedding.engine.renderer.OnFirstFrameRenderedListener;
+
+/**
+ * Interface between Flutter embedding's Java code and Flutter engine's C/C++ code.
+ *
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ *
+ * Flutter's engine is built with C/C++. The Android Flutter embedding is responsible for
+ * coordinating Android OS events and app user interactions with the C/C++ engine. Such coordination
+ * requires messaging from an Android app in Java code to the C/C++ engine code. This
+ * communication requires a JNI (Java Native Interface) API to cross the Java/native boundary.
+ *
+ * The entirety of Flutter's JNI API is codified in {@code FlutterJNI}. There are multiple reasons
+ * that all such calls are centralized in one class. First, JNI calls are inherently static and
+ * contain no Java implementation, therefore there is little reason to associate calls with different
+ * classes. Second, every JNI call must be registered in C/C++ code and this registration becomes
+ * more complicated with every additional Java class that contains JNI calls. Third, most Android
+ * developers are not familiar with native development or JNI intricacies, therefore it is in the
+ * interest of future maintenance to reduce the API surface that includes JNI declarations. Thus,
+ * all Flutter JNI calls are centralized in {@code FlutterJNI}.
+ *
+ * Despite the fact that individual JNI calls are inherently static, there is state that exists
+ * within {@code FlutterJNI}. Most calls within {@code FlutterJNI} correspond to a specific
+ * "platform view", of which there may be many. Therefore, each {@code FlutterJNI} instance holds
+ * onto a "native platform view ID" after {@link #attachToNative(boolean)}, which is shared with
+ * the native C/C++ engine code. That ID is passed to every platform-view-specific native method.
+ * ID management is handled within {@code FlutterJNI} so that developers don't have to hold onto
+ * that ID.
+ *
+ * To connect part of an Android app to Flutter's C/C++ engine, instantiate a {@code FlutterJNI} and
+ * then attach it to the native side:
```

Comment:
Nothing to do, here but thank you for adding detailed doc comments that give the rationale as well as describe the class.